File: day3.py
#!/usr/bin/env python3
from functools import reduce
from utils import runner

def helper(lines, dx, dy):
    x,y = (0,0)
    count = 0
    while y < len(lines)-1:
        x = x + dx
        y += dy
        if lines[y][x % len(lines[0])] == '#':
            count += 1
    return count

def part1(lines):
    return helper(lines, 3, 1)

def part2(lines):
    slopes = [(1,1), (3,1), (5,1), (7,1), (1,2)]
    [helper(lines, x, y) for (x,y) in slopes]
    return reduce(lambda x,y:x*y, [helper(lines, x, y) for (x,y) in slopes])

if __name__ == "__main__":
    sample = ['..##.......',
              '#...#...#..',
              '.#....#..#.',
              '..#.#...#.#',
              '.#...##..#.',
              '..#.##.....',
              '.#.#.#....#',
              '.#........#',
              '#.##...#...',
              '#...##....#',
              ".#..#...#.#"]
    with open("day3.txt") as f:
        lines = f.readlines()
        lines = [line.strip() for line in lines]
    runner(part1, sample, 7, lines, 181)
    runner(part2, sample, 336, lines, 1260601650)

File: src/day3.rs
use std::fs;

fn helper(lines: &[&str], dx: usize, dy: usize) -> i32 {
    let mut x = 0;
    let mut y = 0;
    let mut count = 0;
    while y < lines.len()-1 {
        x += dx;
        y += dy;
        if lines[y].chars().nth(x % lines[0].len()).unwrap() == '#' {
            count += 1
        }
    }
    count
}

fn part1(input: &[&str]) -> i32 {
    helper(input, 3, 1)
}

fn part2(input: &[&str]) -> i32 {
    let slopes = [[1,1], [3,1], [5,1], [7,1], [1,2]];
    let mut product = 1;
    for slope in slopes.iter() {
        product *= helper(input, slope[0], slope[1]);
    }
    product
}

fn runner(func: &dyn Fn(&[&str]) -> i32, sample: &[&str], expected: i32, input: &[&str], answer: Option<i32>) {
    assert_eq!(func(sample), expected);
    let ans = func(input);
    println!("{}", ans);
    if let Some(x) = answer {
        assert_eq!(ans, x);
    }
}

pub fn day3() {
    let sample = ["..##.......",
                  "#...#...#..",
                  ".#....#..#.",
                  "..#.#...#.#",
                  ".#...##..#.",
                  "..#.##.....",
                  ".#.#.#....#",
                  ".#........#",
                  "#.##...#...",
                  "#...##....#",
                  ".#..#...#.#"].to_vec();
    let contents = fs::read_to_string("day3.txt").unwrap();
    let input: Vec<&str> = contents.lines().collect();
    runner(&part1, &sample, 7, &input, Some(181));
    runner(&part2, &sample, 336, &input, Some(1260601650));
}